### DISCUSSION TITLE
Encrypted gossipsub with set destinations

### DISCUSSION THREAD

**Discussion by Darin755** [libp2p/specs#655] [Ideas]
Is it possible to possibly add some sort of group encryption to gossipsub? What I am looking to do allow one node to send data to a bunch of other nodes but not to the public. My idea is that there could be some sort of way to specify the resipiants when sending a gossipsub message. This would be enforced by encrypting the message so that it can be decrypted by a specified list of peers. These messages could still be routed like normal but sensitive data would remain safe.

From a cryptography perspective it would be hard as you would need to have one message that could be decrypted by many peers. I think the easiest way to do this would to have a decryption key that is encrypted for each receiving node.

**Comment by vyzo** [comment]
From a p2p perspective it also is dodgy: why would you blindly forward messages you cant possibly validate?

**Reply by Darin755** [reply]
Currently does gossipsub have any kind validity check? What happens if a node tampers with a message?

With encryption a message could be dropped if it is tampered with or corrupted.

**Reply by vyzo** [reply]
Yes, validation is a very important part of building a secure network, but you have to BYO.

**Reply by vyzo** [reply]
The point is that unless you carry some sort of zk proof, you cant reasonably forward those messages.
